starters/express-typescript-typeorm-postgres/src/controllers/technology/handlers/get-technology.ts

```diff
@@ -1,50 +1,63 @@
 import { NextFunction, Request, Response } from 'express';
 import { getReasonPhrase, StatusCodes } from 'http-status-codes';
+import { useCache } from '../../../cache/cache';
 import { Result } from '../../../constants/result';
 import { dataSource } from '../../../db/datasource';
 import { Technology } from '../../../entities/technology.entity';
-import { ErrorResult, SuccessResult } from '../../../interfaces/results';
+import { ErrorResult, NotFoundResult, SuccessResult } from '../../../interfaces/results';
+import { LogHelper } from '../../../utils/log-helper';
+
+type TechnologyResult = SuccessResult<Technology> | NotFoundResult | ErrorResult;
 
 export async function getTechnology(
   req: Request,
   res: Response,
   next: NextFunction
 ): Promise<void> {
-  try {
-    const technologyId: number = parseInt(req.params.technologyId);
-    const technologyResult = await findTechnology(technologyId);
+  const technologyId: number = parseInt(req.params.technologyId);
+  const technologyResult = await useCache(req.originalUrl, () => findTechnology(technologyId));
+
+  if (technologyResult.type === Result.ERROR) {
+    LogHelper.error(technologyResult.message, technologyResult.error);
+    next(technologyResult.error);
+    return;
+  }
 
-    if (technologyResult.type === Result.ERROR) {
-      res.status(StatusCodes.NOT_FOUND).json({
-        error: getReasonPhrase(StatusCodes.NOT_FOUND),
-        details: technologyResult.message,
-      });
-      return;
-    }
-    res.json(technologyResult.data);
-  } catch (e) {
-    next(e);
+  if (technologyResult.type === Result.NOT_FOUND) {
+    res.status(StatusCodes.NOT_FOUND).json({
```

Comment:
My reasoning behind creating a new Result for `NOT_FOUND` is that in this scenario we have the second if under the first one, and I think that is more readable.

My problem of having different error types is that it would then need if cases inside an if case, which increases cognitive complexity

Who knows maybe in the future there are going to be more result types, like BAD_REQUEST, or FORBIDDEN. You know I'm not sure either anymore. 🤔

Reply:
🤷